File: src/main/java/jila/Window.java
package jila;

import org.lwjgl.Version;
import org.lwjgl.glfw.GLFWErrorCallback;
import org.lwjgl.opengl.GL;

import static org.lwjgl.glfw.Callbacks.glfwFreeCallbacks;
import static org.lwjgl.glfw.GLFW.*;
import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.system.MemoryUtil.NULL;

/**
 * Created by Hojat Ghasemi on 2022-02-04.
 * The author could be contacted at "https://twitter.com/hojat_93"
 */
public class Window {
    private int width, height;
    private String title;
    private long glfwWindow;

    private static Window window;

    private Window() {
        this.width = 1920;
        this.height = 1080;
        this.title = "Mario";
    }

    /**
     * this class is singleton.
     *
     * @return only one instance of Window will be created.
     */
    public static Window get() {
        if (Window.window == null) {
            Window.window = new Window();
        }
        return Window.window;
    }

    public void run() {
        //https://www.lwjgl.org/guide

        System.out.println("Hello LWJGL " + Version.getVersion() + "!");

        init();
        loop();

        // Free up the memory after user closes the window
        glfwFreeCallbacks(glfwWindow);
        glfwDestroyWindow(glfwWindow);

        // Terminate GLFW
        glfwTerminate();
        glfwSetErrorCallback(null).free();

    }

    public void init() {
        // Setting up an error callback
        GLFWErrorCallback.createPrint(System.err).set();

        // Initialize GLFW
        if (!glfwInit()) {
            throw new IllegalStateException("Unable to initialize GLFW.");
        }

        // Configure GLFW
        glfwDefaultWindowHints();
        glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE);
        glfwWindowHint(GLFW_RESIZABLE, GLFW_TRUE);
        glfwWindowHint(GLFW_MAXIMIZED, GLFW_TRUE);

        // Create the window
        glfwWindow = glfwCreateWindow(this.width, this.height, this.title, NULL, NULL);
        if (glfwWindow == NULL) {
            throw new IllegalStateException("Failed to create the GLFW window.");
        }

        // these callbacks are defined in MouseListener.java and are fed to GLFW in here, after the window is created.
        glfwSetCursorPosCallback(glfwWindow, MouseListener::mousePosCallback);
        glfwSetMouseButtonCallback(glfwWindow, MouseListener::mouseButtonCallback);
        glfwSetScrollCallback(glfwWindow, MouseListener::mouseScrollCallback);
        // Adding the keyboard callback
        glfwSetKeyCallback(glfwWindow, KeyListener::keyCallback);


        // Make the OpenGL context current.
        glfwMakeContextCurrent(glfwWindow);
        // Enable v-sync
        glfwSwapInterval(1);

        // Make the window visible
        glfwShowWindow(glfwWindow);

        // This line is critical for LWJGL's interoperation with GLFW's
        // OpenGL context, or any context that is managed externally.
        // LWJGL detects the context that is current in the current thread,
        // creates the GLCapabilities instance and makes the OpenGL
        // bindings available for use.
        GL.createCapabilities();
    }

    public void loop() {
        while (!glfwWindowShouldClose(glfwWindow)) {
            // Poll events
            glfwPollEvents();

            glClearColor(1f, 1f, 1f, 1f);
            glClear(GL_COLOR_BUFFER_BIT);

            // testing if our key callback works correctly.
            if (KeyListener.isKeyPressed(GLFW_KEY_SPACE)) {
                System.out.println("Imagine I'm jumping!");
            }

            glfwSwapBuffers(glfwWindow);
        }


    }


}

File: src/main/kotlin/jila/Window.kt
package jila

import org.lwjgl.Version
import org.lwjgl.glfw.GLFW.*
import org.lwjgl.glfw.GLFWErrorCallback
import org.lwjgl.opengl.GL
import org.lwjgl.opengl.GL11.glClearColor
import org.lwjgl.system.MemoryUtil.NULL

/**
 * Created by Hojat Ghasemi on 2022-01-31.
 *The author could be contacted at "https://twitter.com/hojat_93"
 */
object Window {
    private var width = 1920
    private var height = 1080
    private var title = "Mario"
    private var glfwWindow: Long = 0

    fun run() {
        println("Hello LWJGL ${Version.getVersion()}!")
        init()
        loop()
    }

    private fun loop() {

        while (!glfwWindowShouldClose(glfwWindow)) {
            // Poll events
            glfwPollEvents()
            glClearColor(1f, 0f, 0f, 1f)
        }

    }

    private fun init() {
        // Setup error callback.
        GLFWErrorCallback.createPrint(System.err).set()

        // Initialize GLFW
        if (glfwInit()) {
            throw IllegalStateException("Unable to initialize GLFW.")
        }

        // Configure GLFW
        glfwDefaultWindowHints()
        glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE)
        glfwWindowHint(GLFW_RESIZABLE, GLFW_TRUE)
        glfwWindowHint(GLFW_MAXIMIZED, GLFW_TRUE)

        // Create the window
        glfwWindow = glfwCreateWindow(width, height, title, NULL, NULL)
        if (glfwWindow == NULL) {
            throw java.lang.IllegalStateException("Failed to create the GLFW window.")
        }

        // Make the OpenGL context current
        glfwMakeContextCurrent(glfwWindow)

        // Enable v-sync (there will be no delay between frames)
        glfwSwapInterval(1)

        // Show the window
        glfwShowWindow(glfwWindow)

        /*
        * This line is critical for LWJGL's interoperation with GLFW's OpenGL context, or any context that is
        * managed externally. LWJGL detects the context that is current in the current thread, creates the
        * GLCapabilities instance and makes the OpenGL bindings available for use.*/
        GL.createCapabilities()

    }
}